src/collapse/vsprof.rs

```diff
@@ -0,0 +1,288 @@
+                    // `A()`. This leaves `A()` with the correct number of 500 samples.
+                    //
+                    // The top node is always written, but while walking down the stack, if the
+                    // previous number of calls is equal to the current number of calls, we don't
+                    // want to write the current top node, because that would duplicate the number
+                    // of samples for the current node.
+                    let mut prev_number_of_calls = 0;
+                    for _ in 0..(prev_depth - depth + 1) {
+                        if prev_number_of_calls != self.stack.last().unwrap().1 {
+                            self.write_stack(occurences);
+                        }
+                        prev_number_of_calls = self.stack.pop().unwrap().1;
+
+                        if self.stack.is_empty() {
+                            break;
+                        }
+
+                        let last = self.stack.len() - 1;
+                        let number_of_calls = &self.stack[last].1;
+                        if prev_number_of_calls < *number_of_calls {
+                            self.stack[last].1 -= prev_number_of_calls;
+                        }
+                    }
+
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+            }
+        } else {
+            return invalid_data_error!("Unable to parse function name from line:\n{}", line);
+        }
+
+        Ok(())
+    }
+
+    // Store the current stack in `occurences`
+    fn write_stack(&self, occurrences: &mut Occurrences) {
+        if let Some(nsamples) = self.stack.last().map(|(_, n)| *n).filter(|n| *n > 0) {
+            let functions: Vec<_> = self.stack.iter().map(|(f, _)| &f[..]).collect();
+            occurrences.insert(functions.join(";"), nsamples);
+        }
+    }
+}
+
+/// Gets the number from the start of the line. This can either be a number <1000, in which case the
+/// line doesn't contain double quotes, or the number can be >1000, in which case the line does
+/// contain double quotes. In both cases `line` may start with a leading comma, which will be
+/// ignored.
+///
+/// ### Example inputs
+/// - Number <1000: `471,91.25,18.39,401.92,81.02,"Raytracer.exe",`
+/// - Number >1000: `"2,893,824",54.37,4.21,0.04,0.00,"Raytracer.exe",`
+fn get_next_number(line: &str) -> io::Result<(usize, &str)> {
+    // Trim the leading comma, if any
+    let line = line.strip_prefix(',').unwrap_or(line);
+
+    let mut remove_leading_comma = false;
+    let field = if let Some(line) = line.strip_prefix('"') {
+        remove_leading_comma = true;
+        line.split_once('"')
+    } else {
+        line.split_once(',')
+    };
+
+    if let Some((num, remainder)) = field {
+        // Parse the number
+        let thousands = num.split(|c: char| !c.is_ascii_digit());
```

Comment:
Oh, here's maybe one more. Should this perhaps read
```suggestion
        let thousands = num.split(|c| c == ',');
```

Reply:
Wouldn't that then present a problem with different output cultures? I'm not sure whether Visual Studio always outputs the numbers in the `en-US` culture (so with `,` as thousands separators and `.` as the decimal symbol), or whether they'll output using the current culture, which may use `,` and `.` in the opposite way (so `.` as thousands separator and `,` as the decimal symbol).

To handle that, and to be a bit more explicit, maybe this would be more appropriate:
```suggestion
        let thousands = num.split(|c: char| c == ',' || c == '.');
```

Reply:
Hmm, it's an interesting point. What I was trying to get at was that there are some characters we probably shouldn't be splitting on because it can make errors worse (e.g., imagine if we get a field that's a word not a number). I think I'd be on board with splitting on `,`, `.`, and ` `.

Reply:
Why also split on ` `? A space would never occur in an actual number, so why would we split on that?

To get a better error message when the field is a word instead of a number, should we perhaps verify that the field only contains digits, `.`s and `,`s? Although that might introduce some (significant) overhead, as `get_next_number` is called twice for every line in the profile output (once to get the depth and once to get the number of calls), and a large profile could have many lines.

Right now a word instead of a number returns `Expected thousands separators for number bigger than 1000, found '{}'`, which is not exactly a great error.

Reply:
Space-separated thousands is actually [fairly common](https://docs.oracle.com/cd/E19455-01/806-0169/overview-9/index.html).

Here's a counter-proposal: parse one digit (char) at a time. So instead of going by thousands, we go by tens. At each character:
- if it's a digit, `*= 10; += c - '0';`
- if it's `,`, `.`, or ` `, skip
- otherwise error out and print the whole field

How does that sound?

Reply:
> Space-separated thousands is actually [fairly common](https://docs.oracle.com/cd/E19455-01/806-0169/overview-9/index.html).

Huh, you learn something new every day I guess haha.

Your proposal sounds good, when I have some time this week I'll give that a shot.